### ISSUE TITLE
[TOKEN_EXPIRED] App loses auth state on user registration and page refresh

### ISSUE THREAD

**Issue by priyath** [firebase/firebase-js-sdk#6175]
## Describe your environment

  * Operating System version: macOS Big Sur Version 11.4
  * Browser version: Google Chrome Version 100.0.4896.127 (Official Build) (arm64)
  * Firebase SDK version: 9.6.11
  * Firebase Product: auth

## Describe the problem
When a new user (email is not yet registered with firebase authentication) signs in using Google Login, and the app
is refreshed immediately after (approx 10 seconds), the authenticated state is lost. On closer inspection, it was identified that a  network call initiated by the firebase SDK upon page refresh (https://identitytoolkit.googleapis.com/v1/accounts:lookup?key=<key>) errors out with a 400 TOKEN_EXPIRED error message.

```
{
  "error": {
    "code": 400,
    "message": "TOKEN_EXPIRED",
    "errors": [
      {
        "message": "TOKEN_EXPIRED",
        "domain": "global",
        "reason": "invalid"
      }
    ]
  }
}
```

## Steps to reproduce:
A basic working example of this can be found at https://github.com/priyath/firebase-auth-issue

1. Checkout the above repository
2. Update firebase.js file with your project's firebase credentials
3. Run the application using yarn start
4. Open the browser console and click on Sign Up. Sign Up using credentials of an account that is not yet registered with your firebase project.
5. Observe both the console and the network tab.

## Observations:
Upon successful login (and register) The auth state changed log will fire with the User object. 
After 10 seconds, a page refresh will be triggered, after which the auth state changed log will be null.
If you inspect the network tab, the request to https://identitytoolkit.googleapis.com/v1/accounts:lookup?key=<key> will show a 400 TOKEN_EXPIRED error.

**Comment by Paul-Todd** [comment]
I am also seeing a very similar issue with 9.7.0 (but also in 9.8.0)

this works:
`
    this.firebase = initializeApp(firebaseConfig);
    this.auth = getAuth(this.firebase);
`

but this does not
`
this.firebase = initializeApp(firebaseConfig);
    this.auth = initializeAuth(this.firebase, { errorMap: debugErrorMap });
`

Right now I have rolled back to the old code until I have more chance to look at what is going wrong - no errors and no changes in the execution flow were noted in my app.

I also did the tokenRefresh result to no effect

**Comment by sam-gc** [comment]
@Paul-Todd, the issue with the second example is that you're not providing any persistence type in the options map. [`getAuth()`](https://firebase.google.com/docs/reference/js/auth.md#getauth) includes a bunch of automatic dependencies (like multiple types of persistence), whereas [`initializeAuth`](https://firebase.google.com/docs/reference/js/auth.md#initializeauth) allows you to tweak your dependencies (but that means you need to explicitly include a `persistence` field, or else Auth defaults to in-memory).

Please see https://firebase.google.com/docs/auth/web/custom-dependencies for more information about `getAuth` vs `initializeAuth`.

The problem in this issue is different, I believe, so @priyath please respond to the questions in https://github.com/firebase/firebase-js-sdk/issues/6175#issuecomment-1115099159 when you get the chance.

**Comment by priyath** [comment]
@sam-gc apologies for the delay!

Immediately after sign up the indexDB has the following entry:
<img width="811" alt="Screenshot 2022-05-10 at 10 53 10" src="https://user-images.githubusercontent.com/10166923/167553041-04055aba-a6e5-4ad9-8e26-8841a547e278.png">

After a 10 second page refresh: (this can even be a manual refresh after a few seconds has elapsed as well.)
<img width="807" alt="Screenshot 2022-05-10 at 10 53 44" src="https://user-images.githubusercontent.com/10166923/167553501-d13e9e1b-e277-47b1-9a0c-f384c89d448f.png">

With `user.getIdToken(/*forceRefresh*/ true))` 

If I immediately perform `user.getIdToken(/*forceRefresh*/ true))` after sign up, the token refreshes successfully and works as expected.

However, if I delay this by a few seconds (~ 10 seconds), the token refresh results in a token expired API response.
<img width="1045" alt="Screenshot 2022-05-10 at 11 16 48" src="https://user-images.githubusercontent.com/10166923/167557393-a7026bdc-27df-49a0-abb1-db16d653817e.png">

I wonder if there could be something in my firebase project that is causing this behavior? If not, I can also go with a forced refresh immediately after user registration which seems like a possible workaround for this problem, although it would still be bit of a race condition between token expiration and the token refresh events.

**Comment by kefniark** [comment]
I run into a really similar issue, some of my users complained about having to login twice.

I found out during the first login (user creation), the token I get from firebase seem to have indeed a really short lifetime and at one of the first query to my backend, it get refused (less than 1-2s after login)
```ts
// this throw an error with `FirebaseAuthError: Firebase ID token has expired. `
await firebase.auth().verifyIdToken(token)
```

What I noticed:
- It only seem to happen for me when a user is created on firebase side, the first time. Any later login go through as expected.
- I can reproduce the issue consistently by going to firebase admin, delete a user, login with with that user again
- In my case, I suspect it to be related to `firebase.auth().setCustomUserClaims()` usage

Possible workaround:
- Adding a delay `setTimeout(login, 500)` on the frontend is enough for `auth.currentUser?.getIdToken()` to become valid, but this is ugly and doesn't feel reliable
- Forcing a frontend token refresh just after user creation `await auth.currentUser?.getIdTokenResult(true)` like you suggested in your previous comment

I hope this may help you
